test/bpf_tests/bpf_test.go

```diff
@@ -0,0 +1,470 @@
+	"regexp"
+	"strconv"
+	"strings"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/perf"
+	"github.com/golang/protobuf/proto"
+	"github.com/vishvananda/netlink/nl"
+	"google.golang.org/protobuf/encoding/protowire"
+
+	"github.com/cilium/cilium/pkg/byteorder"
+	"github.com/cilium/cilium/pkg/datapath/link"
+	"github.com/cilium/cilium/pkg/monitor"
+)
+
+var testPath = flag.String("bpf-test-path", "", "Path to the eBPF tests")
+
+func TestBPF(t *testing.T) {
+	if testPath == nil || *testPath == "" {
+		t.Fatal("-bpf-test-path is a required flag")
+	}
+
+	entries, err := os.ReadDir(*testPath)
+	if err != nil {
+		t.Fatal("os readdir: ", err)
+	}
+
+	pinPath := "/sys/fs/bpf/cilium-test"
+	defer func() {
+		os.RemoveAll(pinPath)
+	}()
+	if err = os.Mkdir(pinPath, 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	for _, entry := range entries {
+		if entry.IsDir() {
+			continue
+		}
+
+		if !strings.HasSuffix(entry.Name(), ".o") {
+			continue
+		}
+
+		loadAndRunSpec(t, entry, pinPath)
+	}
+}
+
+func loadAndRunSpec(t *testing.T, entry fs.DirEntry, pinPath string) {
+	elfPath := path.Join(*testPath, entry.Name())
+	spec := loadAndPrepSpec(t, elfPath)
+
+	const maxLoadRetries = 5
+	var (
+		coll    *ebpf.Collection
+		err     error
+		logSize = 1 << 20 // 1 MiB
+	)
+	for i := 0; i < maxLoadRetries; i++ {
+		coll, err = ebpf.NewCollectionWithOptions(spec, ebpf.CollectionOptions{
+			Programs: ebpf.ProgramOptions{
+				LogSize: logSize,
+			},
+			Maps: ebpf.MapOptions{
+				PinPath: pinPath,
+			},
+		})
+		if err != nil {
+			// Use this definition of ENOSPC instead of syscall.ENOSPC, since the stdlib constant is only
+			// available on linux. Even if this doesn't run, we still want it to compile on non-linux systems.
+			const ENOSPC = syscall.Errno(0x1c)
+			// ENOSPC usually means that the log size was to small, so increase and retry.
+			if errors.Is(err, ENOSPC) {
+				// 1, 4, 16, 64, 256 MiB
+				logSize = logSize << 2
+				continue
+			}
+
+			t.Fatal("new coll:", err)
+		}
+
+		break
+	}
+	defer coll.Close()
+
+	err = loadCallsMap(spec, coll)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testNameToPrograms := make(map[string]programSet)
+
+	for progName, spec := range spec.Programs {
+		match := checkProgRegex.FindStringSubmatch(spec.SectionName)
+		if len(match) == 0 {
+			continue
+		}
+
+		progs := testNameToPrograms[match[1]]
+		if match[2] == "setup" {
+			progs.setupProg = coll.Programs[progName]
+		}
+		if match[2] == "check" {
+			progs.checkProg = coll.Programs[progName]
+		}
+		testNameToPrograms[match[1]] = progs
+	}
+
+	for progName, set := range testNameToPrograms {
+		if set.checkProg == nil {
+			t.Fatalf(
+				"File '%s' contains a setup program in section '%s' but no check program.",
+				elfPath,
+				spec.Programs[progName].SectionName,
+			)
+		}
+	}
+
+	// Collect debug events and add them as logs of the main test
+	var globalLogReader *perf.Reader
+	if coll.Maps["test_cilium_events"] != nil {
```

Comment:
For readability:
```suggestion
	if m := coll.Maps["test_cilium_events"]; m != nil {
```
and `perf.NewReader(m, ...)`